### PR TITLE
Integration test

### PR DESCRIPTION
**Pull request by AHP15**
In this PR, I added the following:

 - integration tests for:
    - User index page
    - user show page
    - User post index page
    - Post show page
 - N+1 problem solved
  before:

before_last

After:

after_last

Cheers 👍

### CODE REVIEW

File: spec/integration/user/index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      expect(page).to have_content('Number of posts:')
+    end
+
+    it 'When I click on a user, I am redirected to that user`s show page' do
+      visit '/'
+      click_link 'test1'
+      expect(page).to have_content('Bio')
+    end
+  end
+end
```
Comment (TedLivist) [inline spec/integration/user/index_spec.rb]: - [x] Kindly add tests that assert we can see the user's username
- [x] Kindly add an assertion that demonstrates when a user is clicked the page redirects to the user's show page.

File: spec/integration/user/show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      visit '/users/108'
+      expect(page).to have_content('Bio')
+    end
+
+    it 'I can see the user`s last 3 posts' do
+      visit '/users/108'
+      expect(page).to have_selector('a')
+      expect(page).to have_selector('a')
+      expect(page).to have_selector('a')
+    end
```
Comment (TedLivist) [inline spec/integration/user/show_spec.rb]: - [x] Kindly rewrite this test for checking the first three posts and ensure it is more thorough. There is no indication that these assertions list the three posts. You are just selecting any item on the page with the "a" tag which can be anything really. I would recommend you use the actual content of the posts themselves like their individual post titles.

File: spec/integration/user/show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    it 'I can see a button that lets me view all of a user`s posts.' do
+      visit '/users/108'
+      expect(page).to have_content('See all posts')
+    end
+  end
+end
```
Comment (TedLivist) [inline spec/integration/user/show_spec.rb]: - [x] Kindly add assertions for the following scenarios:

  - Tests that redirect to the post's show page when you click on a user's post.
  - Tests that redirect to the user's post's index page when you click on a see all posts.

File: spec/integration/post/index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  it 'I can see how many comments a post has.' do
+    visit '/users/108/posts'
+    expect(page).to have_content('Comments:')
+  end
+
+  it 'I can see how many likes a post has.' do
+    visit '/users/108/posts'
+    expect(page).to have_content('Likes:')
+  end
+end
```
Comment (TedLivist) [inline spec/integration/post/index_spec.rb]: - [x] Kindly add a test that redirects to the post's show page when you click on a post.

File: spec/integration/post/show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  it 'I can see how many comments a post has.' do
+    visit '/users/108/posts/48'
+    expect(page).to have_content('Comments:')
+  end
+
+  it 'I can see how many likes a post has.' do
+    visit '/users/108/posts/48'
+    expect(page).to have_content('Likes:')
+  end
+end
```
Comment (TedLivist) [inline spec/integration/post/show_spec.rb]: - [x] Kindly add:

  - Tests that assert you can see the username of each commenter.
  - Tests that assert you can see the comment each commentor left.

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @Abdessittir and @Kweeka1,

Good job so far!
✔️ Modularization of tests
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
- [x] All your tests are failing. Kindly note that by convention, all the variables and data needed for running tests should be created within the test environment (or passed from an external environment if applicable). I can see from your test URLs that you visit users with the ID of 108 and post ID of 48, but since there are no users and posts with those IDs in your test environment your tests will fail outside your PC. Given the same arguments tests should be able to run pass regardless of the system it runs on which is why I would recommend you create your user, posts, comments, and likes within the test files and then visit the routes using the `user.id` and `post.id` instead of `108`, and `48`.

  ruie

_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ichala — APPROVED** [review]
Hi Student  ,
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:  
 
68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f6c3366516333614356304871707a5763552f67697068792e676966

# To Highlight 💪🏻

- Changes Implemented ✔️ 


 Good job implementing the requirements  :+1: Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Optional Suggestions  :

- N/A


**********
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag  ichala in your question so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**